go/internal/flag/flag.go

```diff
@@ -0,0 +1,156 @@
+/*
+Copyright 2022 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package flag is an internal package to allow us to gracefully transition
+// from the standard library's flag package to pflag. See VEP-4 for details.
+//
+// In general, this package should not be imported or depended on, except in the
+// cases of package servenv, and entrypoints in go/cmd. This package WILL be
+// deleted after the migration to pflag is completed, without any support for
+// compatibility.
+package flag
+
+import (
+	goflag "flag"
+	"fmt"
+	"os"
+	"reflect"
+	"strings"
+
+	"vitess.io/vitess/go/vt/log"
+)
+
+// Parse wraps the standard library's flag.Parse to perform some sanity checking
+// and issue deprecation warnings in advance of our move to pflag.
+//
+// It also adjusts the global CommandLine's Usage func to print out flags with
+// double-dashes when a user requests the help, attempting to otherwise leave
+// the default Usage formatting unchanged.
+//
+// See VEP-4, phase 1 for details: https://github.com/vitessio/enhancements/blob/c766ea905e55409cddeb666d6073cd2ac4c9783e/veps/vep-4.md#phase-1-preparation
+func Parse() {
+	// First, override the Usage func to make flags show in their double-dash
+	// forms to the user.
+	goflag.CommandLine.Usage = func() {
+		fmt.Fprintf(goflag.CommandLine.Output(), "Usage of %s:\n", os.Args[0])
+
+		var buf strings.Builder
+		goflag.CommandLine.VisitAll(func(f *goflag.Flag) {
+			defer buf.Reset()
+			defer func() { fmt.Fprintf(goflag.CommandLine.Output(), "%s\n", buf.String()) }()
```

Comment:
we have to wrap this defer inside an anonymous function to delay the execution of `buf.String()`

the other thing i briefly considered was:

```go
defer buf.Reset()

...

if bf, ok := f.Value.(maybeBoolFlag); ok && ... {
    fmt.Fprintf(...)
    goto print
}

// print first and second lines for non-bool flags to `buf`

print:
    fmt.Fprintf(goflag.CommandLine.Output(), "%s\n", buf.String())
```

but ended up deciding the goto was more gross than the `defer func() { ... }()`

Reply:
IMO, the `defer` does not seem bad. Another alternative would be to create an anonymous function where we have the current `defer`, and call that function where we `return`.